report/3.Theory/Parser.tex

```diff
@@ -0,0 +1,112 @@
+
+\subsection{Output data types}
+
+The parser will output some data type, that represents the compiled program in a hierarchical structure. 
+The outputted data type should hold all the information or almost all the information needed to generate the code later. 
+Sometimes the data type will be accommodated by a symbol table in order to generate the desired code.
+
+\textbf{Parse tree}
+
+One of the possible data structures to be outputted from the parser is the parse tree explained in section \ref{CFGdescription}. 
+The parse tree should resemble the structure of the CFG and the concrete input to the parser. 
+The parse tree is easily generated, and often the actual output from the parser. 
+However, the parse tree contains a lot of excessive information, which is not needed to understand the complete meaning of the program. 
+This means that a parse tree is both hard to analyze and slow to traverse. 
+Therefore, another common output type from the parser is the abstract syntax tree.
+
+\textbf{Abstract syntax tree}
+
+The abstract syntax tree is a sparse version of the parse tree. 
+It still contains all the information needed to understand the structure of the program, but in a more compressed way. 
+For instance parentheses can be omitted in the abstract syntax tree.
+
+Recall the parse tree from figure \ref{fig:AmbiguousGrammarEx1} a corresponding abstract syntax tree can bee seen on figure \ref{fig:ast} 
+
+\begin{figure}[H]
+	\centering
+	\includegraphics[width=\textwidth/2+\textwidth/4]{3.Theory/images/ast.png}
+	\caption{
+		Abstract syntax tree for parse tree in figure \ref{fig:AmbiguousGrammarEx1}.
+	}
+	\label{fig:ast}
+\end{figure}
+
+%\subsection{Grammar}
+%
+%The parser is built upon the CFG defined for the language. 
+%It is the responsibility of the CFG to tell the parser how to parse the language. 
+%For instance, when writing a top down parser, first and follow sets can be generated on behalf of the CFG. 
+%These will help determine what should be parsed next. 
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% I guess it is not really necessary to talk about CFG, when we already have a big section about it.
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+\subsection{Contextual analysis}
+
+Apart from syntactical analysis the parser is also concerned with contextual analysis. 
+The contextual analysis ensures that types are correct, immutable variables are not reassigned and other contextual things that cannot be inferred from the CFG due to its context free nature. 
+The major part of the contextual analysis is the type checking.
+
+\textbf{Type checking}
+
+The type checking part of the parser often utilizes a symbol table.
+
+INSERT JONAS SYMBOL TABLE STUFF HERE.
+
+It is also the responsibility of this part of the parser to provide any relevant type information that may be needed in order to proceed with compilation in later phases. 
```

Comment:
Consider using portion or section instead of this "PART" of the parser. (mostly to avoid repetitiveness.